enhancements/network/ingress-logging-api.md

```diff
@@ -0,0 +1,364 @@
+
+For high-traffic clusters, the quantity of log output may exceed the capacity of
+the logging stack.  It also may be the case that the user desires to integrate
+with logging infrastructure that exists outside of OpenShift.  In such
+scenarios, the user may desire that logs go to a custom syslog endpoint.
+
+### Goals
+
+1. Enable the user to configure access logging for an IngressController.
+2. Provide options to send logs to either a sidecar or a custom syslog endpoint.
+3. Provide an option to configure the log format.
+4. Accommodate future additions of possible logging destinations.
+5. Accommodate future additions of logging types besides access logs.
+6. Accommodate future additions of logging parameters.
+7. Avoid redundancy with [the cluster logging log forwarding API](../cluster-logging/cluster-logging-log-forwarding.md).
+
+### Non-Goal
+
+Configuring log collection or aggregation is out of scope.
+
+## Proposal
+
+The IngressController API is extended by adding an optional `Logging` field with
+type `*IngressControllerLogging` to `IngressControllerSpec`:
+
+```go
+type IngressControllerSpec struct {
+	// ...
+
+	// logging defines parameters for what should be logged where.
+	//
+	// +optional
+	Logging *IngressControllerLogging `json:"logging,omitempty"`
+}
+```
+
+The `IngressControllerLogging` type has an optional `Access` field of type
+`*AccessLogging` for configuring access logging; additional fields for other
+types of logging may be added in the future:
+
+```go
+// IngressControllerLogging describes what should be logged where.
+type IngressControllerLogging struct {
+	// access describes how the client requests should be logged.
+	//
+	// If this field is empty, access logging is disabled.
+	//
+	// +optional
+	Access *AccessLogging `json:"access,omitempty"`
+}
+```
+
+The `AccessLogging` type allows specifying a destination using the `Destination`
+field, which has type `LoggingDestination`, and the log format using the
+`HttpLogFormat` field, which has type `string`:
+
+```go
+// AccessLogging describes how client requests should be logged.
+type AccessLogging struct {
+	// destination is where access logs go.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Destination LoggingDestination `json:"destination"`
+
+	// httpLogFormat specifies the format of the log message for an HTTP
+	// request.
+	//
+	// If this field is empty, log messages use the implementation's default
+	// HTTP log format.  For HAProxy's default HTTP log format, see the
+	// HAProxy documentation:
+	// http://cbonte.github.io/haproxy-dconv/1.7/configuration.html#8.2.3
+	//
+	// +optional
+	HttpLogFormat string `json:"httpLogFormat,omitempty"`
+}
+```
+
+The `LoggingDestination` type is a union type:
+
+```go
+// LoggingDestination describes a destination for log messages.
+// +union
+type LoggingDestination struct {
+	// type is the type of destination for logs.  It must be one of the
+	// following:
+	//
+	// * Container
```

Comment:
Are there any additional warnings worth mentioning for this strategy? Like, what are the possible implications of enabling sidecar logging in a high-volume environment?